### PR TITLE
fix(router): ensure subgraph access logs handles null requests/errors

### PR DESCRIPTION
**Pull request by df-wg**
## Motivation and Context

There was a null pointer when consumers would encounter a subgraph with a null request (which can happen in error cases). This PR makes the code much more resilient, and adds tests to ensure that nil pointers are handled gracefully.

Resolves ENG-6173
## Checklist

- [ ] I have discussed my proposed changes in an issue and have received approval to proceed.
- [ ] I have followed the coding standards of the project.
- [ ] Tests or benchmarks have been added or updated.
- [ ] Documentation has been updated on [https://github.com/wundergraph/cosmo-docs](https://github.com/wundergraph/cosmo-docs).
- [ ] I have read the [Contributors Guide](https://github.com/wundergraph/cosmo/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: router/core/engine_loader_hooks.go [hunk 1]
```diff
@@ -141,10 +141,14 @@ func (f *engineLoaderHooks) OnFinished(ctx context.Context, ds resolve.DataSourc
...
-			fields = append(fields, zap.String("url", responseInfo.Request.URL.String()))
+		path := ds.Name
+		if responseInfo.Request != nil {
+			fields = append(fields, f.accessLogger.GetRequestFields(responseInfo, fields)...)
+			if responseInfo.Request.URL != nil {
+				path = responseInfo.Request.URL.Path
+			}
 		}
-		f.accessLogger.WriteRequestLog(responseInfo, fields)
+		f.accessLogger.WriteLog(path, fields)
```
Comment (df-wg) [inline router/core/engine_loader_hooks.go:151]: Ensure that we're writing the logs even in error cases, without a valid request

File: router/core/request_context_fields.go [hunk 2]
```diff
@@ -69,17 +69,20 @@ func NewDurationLogField(val time.Duration, attribute config.CustomAttribute) za
 }
 
 func AccessLogsFieldHandler(attributes []config.CustomAttribute, err any, request *http.Request, responseHeader *http.Header) []zapcore.Field {
-	reqContext := getRequestContext(request.Context())
-
 	resFields := make([]zapcore.Field, 0, len(attributes))
-	resFields = append(resFields, logging.WithRequestID(middleware.GetReqID(request.Context())))
+
+	var reqContext *requestContext
```
Comment (StarpTech) [inline router/core/request_context_fields.go:74]: Please add a test that validates the subgraph logs when the subgraph call errors, timeout etc...

File: router/core/engine_loader_hooks.go [hunk 3]
```diff
@@ -141,10 +141,14 @@ func (f *engineLoaderHooks) OnFinished(ctx context.Context, ds resolve.DataSourc
 			zap.Int("status", responseInfo.StatusCode),
 			zap.Duration("latency", latency),
 		}
-		if responseInfo.Request != nil && responseInfo.Request.URL != nil {
-			fields = append(fields, zap.String("url", responseInfo.Request.URL.String()))
+		path := ds.Name
+		if responseInfo.Request != nil {
```
Comment (StarpTech) [inline router/core/engine_loader_hooks.go:145]: How can Request be nil? I'd expect that the engine always sets it. If not how we do log the request URL? We need to make the request available before we made the request in the engine. Currently, we set everything when request succeed which is not right.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/core/engine_loader_hooks.go (hunk 3 above)
Comment (df-wg) [reply router/core/engine_loader_hooks.go:145]: Good point! Handled in https://github.com/wundergraph/graphql-go-tools/pull/1016

File: router/internal/requestlogger/subgraphlogger.go [hunk 4]
```diff
@@ -43,19 +43,22 @@ func NewSubgraphAccessLogger(logger *zap.Logger, opts SubgraphOptions) *Subgraph
 	}
 }
 
-func (h *SubgraphAccessLogger) WriteRequestLog(respInfo *resolve.ResponseInfo, subgraphFields []zap.Field) {
+func (h *SubgraphAccessLogger) GetRequestFields(respInfo *resolve.ResponseInfo, subgraphFields []zap.Field) []zap.Field {
```
Comment (jensneuse) [inline router/internal/requestlogger/subgraphlogger.go:46]: I recommend to name this function "RequestFields":
https://chatgpt.com/share/675fdb6a-3738-8010-853f-efa3fa7beee8

Comment (df-wg) [reply router/internal/requestlogger/subgraphlogger.go:46]: Thanks for that authoritative source 😆

File: router/internal/requestlogger/subgraphlogger.go [hunk 5]
```diff
@@ -43,19 +43,22 @@ func NewSubgraphAccessLogger(logger *zap.Logger, opts SubgraphOptions) *Subgraph
...
 	}
 
-	if len(subgraphFields) > 0 {
-		fields = append(fields, subgraphFields...)
-	}
-	h.logger.Info(path, fields...)
+	return fields
+}
+
+func (h *SubgraphAccessLogger) Info(message string, fields ...zap.Field) {
```
Comment (df-wg) [inline router/internal/requestlogger/subgraphlogger.go]: Renamed to `Info` to better fit idiom

**Review by StarpTech — APPROVED** [review]
LGTM
